content/blog/2020-10-28-type-based-alias-analysis.md

```diff
@@ -0,0 +1,138 @@
++++
+title = "Type-Based Alias Analysis"
+[extra]
+bio = """
+  Kenneth Li is a first-semester M.Eng. student who is interested in functional
+  programming. He sings bass in the 
+  [Cornell Hangovers](http://www.hangovers.com/).
+"""
+[[extra.authors]]
+name = "Kenneth Li"
++++
+
+Alias analysis allows compilers to determine which pointers may (or must) 
+refer to the same memory location. This is most useful for the purpose of 
+instruction reordering; if a compiler knows that two memory instructions refer 
+to different memory locations, it can switch the order of the instructions. In 
+_Type-Based Alias Analysis_, Diwan, McKinley, and Moss propose and evaluate 
+alias analyses for Modula-3 (a type-safe language with in heritance) that use 
+type information instead of assignment tracking. Each analysis successively 
+refines the previous: _TypeDecl_ uses only type/subtype compatibility, 
+_FieldTypeDecl_ additionally considers field names and other properties, and 
+_SMFieldTypeRefs_ augments the former with a single pass through the 
+instructions to pick up assignment information. The analyses all do a may-alias 
+analysis; all alias pairs calculated are possible aliases, but all other pairs 
+are confirmed to not be aliases.
+
+## Algorithms
+The _TypeDecl_ analysis relies on a simple predicate: two referenced memory 
+access paths `p` and `q` may be aliases if and only if there exists a type that 
+subtypes both the type of `p` and the type of `q`. If such a type did not 
+exist, for example, then if `p` and `q`were aliases, the object they refer to 
+would have two different, incompatible types. _FieldTypeDecl_ takes this 
+further by adding checks for field names, array accesses, and pointer 
+dereferences. For example, paths `p.f` and `q.g` may be aliases only if the 
+field names `f` and `g` are exactly the same and `p` and `q` themselves may be 
+aliases. Additionally, a pointer dereference `q^` may alias a field `p.f` or an 
+array access `p[i]` only if its address was ever taken by the program (since 
+otherwise no pointer to it could exist). Further checks exist for comparisons 
+between pointers and array accesses, which the authors sum up as follows:
+
+1.	Identical [access paths] always alias each other.
+2.	Two qualified expressions may be aliases if they access the same field in 
+potentially the same object.
+3.	A pointer dereference may reference the same location as a qualified or 
+subscripted expression only if their types are compatible and the program may 
+take the address of the qualified or subscripted expression.
+4.	See (3)
+5.	In Modula-3, a subscripted expression cannot alias a qualified expression.
+6.	Two subscripted expressions are aliases if they may subscript the same 
+array. _FieldTypeDecl_ ignores the actual subscripts.
+7.	For all other cases of [access paths], including two pointer dereferences,
+_FieldTypeDecl_ uses _TypeDecl_ to determine aliases.

```

Comment:
I think this list needs some additional context-setting. It's not really clear what the reader is supposed to get out of reading this list of items, or if the order is significant. There is also lots of suprirrisng terminology that it kind of hard to parse without additional context.

Overall, I'm not sure you really need the list itself… is there any way to summarize what role this reasoning plays in the general idea? Doing so might at least make the list possible to understand, and it might even obviate the need for the list altogether.

Reply:
good point, i removed the list and added a summary to indicate the authors' work in determining alias rules between different AP types